src/Source/AbstractPropertySourceIterator.php

```diff
@@ -0,0 +1,177 @@
+{
+    private const DATE_PARTS = [
+        'y' => 'Y',
+        'm' => 'M',
+        'd' => 'D',
+    ];
+    private const TIME_PARTS = [
+        'h' => 'H',
+        'i' => 'M',
+        's' => 'S',
+    ];
+
+    /**
+     * @var IterableResult
+     */
+    protected $iterator;
+
+    /**
+     * @var PropertyAccessor
+     */
+    protected $propertyAccessor;
+
+    /**
+     * @var string default DateTime format
+     */
+    protected $dateTimeFormat;
+
+    /**
+     * @var array<string>
+     */
+    protected $fields = [];
+
+    /**
+     * @param array<string> $fields Fields to export
+     */
+    public function __construct(array $fields, string $dateTimeFormat = 'r')
+    {
+        $this->propertyAccessor = PropertyAccess::createPropertyAccessor();
+        $this->fields = $fields;
+
+        $this->dateTimeFormat = $dateTimeFormat;
+    }
+
+    public function current()
+    {
+        $current = $this->iterator->current();
+
+        return $this->getCurrentData($current);
+    }
+
+    public function next(): void
+    {
+        $this->iterator->next();
+    }
+
+    public function key()
+    {
+        return $this->iterator->key();
+    }
+
+    public function valid(): bool
+    {
+        return $this->iterator->valid();
+    }
+
+    abstract public function rewind(): void;
+
+    public function setDateTimeFormat(string $dateTimeFormat): void
+    {
+        $this->dateTimeFormat = $dateTimeFormat;
+    }
+
+    public function getDateTimeFormat(): string
+    {
+        return $this->dateTimeFormat;
+    }
+
+    /**
+     * NEXT_MAJOR: Change the method visibility to private.
+     *
+     * @return string An ISO8601 duration
+     */
+    public function getDuration(\DateInterval $interval): string
+    {
+        $datePart = '';
+        foreach (self::DATE_PARTS as $datePartAttribute => $datePartAttributeString) {
+            if ($interval->$datePartAttribute !== 0) {
+                $datePart .= $interval->$datePartAttribute.$datePartAttributeString;
+            }
+        }
+
+        $timePart = '';
+        foreach (self::TIME_PARTS as $timePartAttribute => $timePartAttributeString) {
+            if ($interval->$timePartAttribute !== 0) {
+                $timePart .= $interval->$timePartAttribute.$timePartAttributeString;
+            }
+        }
+
+        if ('' === $datePart && '' === $timePart) {
+            return 'P0Y';
+        }
+
+        return 'P'.$datePart.('' !== $timePart ? 'T'.$timePart : '');
+    }
+
+    /**
+     * @param object|array $current
+     */
+    protected function getCurrentData($current): array
+    {
+        $data = [];
+        foreach ($this->fields as $key => $field) {
+            if (\is_string($field)) {
+                $name = \is_string($key) ? $key : $field;
+                $propertyPath = $field;
+            } else {
+                throw new \TypeError('Unsupported field type. Field should be a string.');
+            }
```

Comment:
```suggestion
            if (!\is_string($field)) {
                throw new \TypeError(sprintf(
                    'Field type must be string, %s given.',
                    \is_object($field) ? 'instance of '.\get_class($field) : \gettype($field)
                );
            }

            $name = \is_string($key) ? $key : $field;
            $propertyPath = $field;
```

Reply:
I plan to support another type https://github.com/sonata-project/exporter/pull/341

Reply:
Great. IMO, if we move the exception before there is no need for the `else` clause. When the new type gets support, an `or` condition could be added (`if (!\is_string($field) && !$field instanceof ExportFieldInterface)`).

Reply:
This will be
```
if (\is_string($field) {
    doThings();
else ($field instanceof ExportFieldInterface) {
    doOtherThings();
} else {
    throw new Exception();
}
```